part/iterator.go

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package part
+
+import (
+	"bytes"
+	"sort"
+)
+
+func prefixSearch[T any](root *node[T], key []byte) (*Iterator[T], <-chan struct{}) {
+	this := root
+
+	for {
+		switch {
+		case bytes.Equal(key, this.prefix[:min(len(key), len(this.prefix))]):
+			return newIterator(this), this.watch
+
+		case bytes.HasPrefix(key, this.prefix):
+			key = key[len(this.prefix):]
+			if len(key) == 0 {
+				return newIterator(this), this.watch
+			}
+
+		default:
+			return newIterator[T](nil), root.watch
+		}
+
+		this = this.find(key[0])
+		if this == nil {
+			return newIterator[T](nil), root.watch
+		}
+	}
+}
+
+type Iterator[T any] struct {
+	next [][]*node[T] // the next set of edges to explore
+}
+
+func newIterator[T any](start *node[T]) *Iterator[T] {
+	if start == nil {
+		return &Iterator[T]{nil}
+	}
+	return &Iterator[T]{[][]*node[T]{{start}}}
+}
+
+func lowerbound[T any](start *node[T], key []byte) *Iterator[T] {
+	// The starting edges to explore. This contains all larger nodes encountered
+	// on the path to the node larger or equal to the key.
+	edges := [][]*node[T]{}
+
+	var traverseToMin func(n *node[T])
+	traverseToMin = func(n *node[T]) {
```

Comment:
nit: definition and declaration separate, could be joined?

Reply:
No it can't be since `traverseToMin` calls `traverseToMin`

Reply:
I lifted `traverseToMin` into a top-level function.